### PR TITLE
Cleanup in serialport

### PR DESCRIPTION
**Pull request by dreamer**
I had some cleanup patches waiting in my local branches, so better merge it in before I forget about them :)

Brings back -Wall warnings count a little bit, -Weffc++ count significantly; Coverity and PVS issues should go down as well.

### CODE REVIEW

**Comment by krcroft** [comment]
@dreamer, this sure earns the 'cleanup' label!  I will go over it in detail. 
Regarding testing this, are there some games or terminal programs you can recommend that I can exercise locally? (perhaps running two copies; each connected to each other?).

**Comment by dreamer** [comment]
This code should be utilized by any softmodem connection, so BBS software should exercise these code paths… I tried to do testing with Terminate 5.00… but I am completely lost and have no idea how to use this software - it's incomprehensible - I have no idea if I even installed it correctly (installer throws messages about how "communication is the future" and some new-age'y stuff at me, and then totally fails to communicate how to configure the damn thing). Your comments in #64 and #172 were very helpful, but I am not interested in running this software myself.

The only real code changes are in the first commit: a3144aa0d87da7b221e2a68e7206f9b3fb82b917 - I crafted it carefully to preserve the behaviour and avoid code duplication (although I did not test it).

All the commits after that are standard field initialization / style fixes / effc++ stuff.

**Comment by krcroft** [comment]
Building with Clang and `-t warnmore` gave another 11 type-safety related warnings (listed below), so dropped those in a branched off your branch.. not sure if that's the right approach (regarding how to absorb these in your PR - or should it be a small follow-up PR? - whatever is easiest/quickest)

https://github.com/dreamer/dosbox-staging/compare/po/cleanup-in-serialport-1...kc/cleanup-in-serialport-2

Will review and test later this morning. 

```
misc_util.cpp:56:3: warning: cast from 'Bit8u *' (aka 'unsigned char *') to 'struct _TCPsocketX *' increases required alignment from 1 to 4 [-Wcast-align]
        ((struct _TCPsocketX*)nativetcpstruct)->ready=0;
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
misc_util.cpp:57:3: warning: cast from 'Bit8u *' (aka 'unsigned char *') to 'struct _TCPsocketX *' increases required alignment from 1 to 4 [-Wcast-align]
        ((struct _TCPsocketX*)nativetcpstruct)->sflag=0;
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
misc_util.cpp:58:3: warning: cast from 'Bit8u *' (aka 'unsigned char *') to 'struct _TCPsocketX *' increases required alignment from 1 to 4 [-Wcast-align]
        ((struct _TCPsocketX*)nativetcpstruct)->channel=(SOCKET) platformsocket;
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
misc_util.cpp:63:4: warning: cast from 'Bit8u *' (aka 'unsigned char *') to 'struct _TCPsocketX *' increases required alignment from 1 to 4 [-Wcast-align]
                ((struct _TCPsocketX*)nativetcpstruct)->
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
misc_util.cpp:65:4: warning: cast from 'Bit8u *' (aka 'unsigned char *') to 'struct _TCPsocketX *' increases required alignment from 1 to 4 [-Wcast-align]
                ((struct _TCPsocketX*)nativetcpstruct)->
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
misc_util.cpp:74:4: warning: cast from 'Bit8u *' (aka 'unsigned char *') to 'struct _TCPsocketX *' increases required alignment from 1 to 4 [-Wcast-align]
                ((struct _TCPsocketX*)nativetcpstruct)->
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
misc_util.cpp:76:4: warning: cast from 'Bit8u *' (aka 'unsigned char *') to 'struct _TCPsocketX *' increases required alignment from 1 to 4 [-Wcast-align]
                ((struct _TCPsocketX*)nativetcpstruct)->
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
misc_util.cpp:173:10: warning: implicit conversion changes signedness: 'Bits' (aka 'long') to 'Bitu' (aka 'unsigned long') [-Wsign-conversion]
                        *size=retval;
                             ~^~~~~~
misc_util.cpp:167:47: warning: implicit conversion loses integer precision: 'Bitu' (aka 'unsigned long') to 'int' [-Wshorten-64-to-32]
                Bits retval = SDLNet_TCP_Recv(mysock, data, *size);
                              ~~~~~~~~~~~~~~~               ^~~~~
misc_util.cpp:194:17: warning: implicit conversion changes signedness: 'Bitu' (aka 'unsigned long') to 'Bits' (aka 'long') [-Wsign-conversion]
                } else return retval;
                       ~~~~~~ ^~~~~~
misc_util.cpp:206:36: warning: implicit conversion loses integer precision: 'Bitu' (aka 'unsigned long') to 'int' [-Wshorten-64-to-32]
        if (SDLNet_TCP_Send(mysock, data, bufsize) != (int)bufsize) {
            ~~~~~~~~~~~~~~~               ^~~~~~~
11 warnings generated.
```

**Comment by krcroft** [comment]
> I tried to do testing with Terminate 5.00… but I am completely lost and have no idea how to use it - it's incomprehensible - I have no idea if I even installed it correctly (installer throws messages about how "communication is the future" and some new-age'y stuff at me, and then totally fails to communicate how to configure the damn thing

That cracked me up.. and is so true! Even though I used _"the most powerful terminal software in the world"_ (or whatever it was claimed) for years as a teenager; it was convoluted enough that it still takes my legs out with accidental hotkey strokes and trap doors requiring a quick 'Ctrl-F9' :-)

Fortunately it's thoroughness makes it great at testing DOSBox's softmodem.

**Comment by dreamer** [comment]
I pushed your commits to this review - this will be the the quickest way - I looked through your changes and all seem ok to me :)

**Comment by krcroft** [comment]
Testing passed w/ BBS software including sending and receiving files.

File: src/hardware/serialport/misc_util.cpp [hunk 1]
```diff
@@ -198,97 +167,79 @@ bool TCPClientSocket::ReceiveArray(Bit8u* data, Bitu* size) {
...
-		
-		if(SDLNet_TCP_Send(mysock, sendbuffer, sendbuffersize)!=sendbuffersize) {
-			isopen=false;
-			return false;
-		}
-	}
+	// buffer is full, get rid of it
+	sendbuffer[sendbufferindex] = data;
+	sendbufferindex = 0;
+	return SendArray(sendbuffer, sendbuffersize);
```
Comment (krcroft) [inline src/hardware/serialport/misc_util.cpp]: Much more readable with the break-out to SendArray (in multiple functions).

**Comment by dreamer** [comment]
Thanks :)

**Comment by krcroft** [comment]
> Coverity and PVS issues should go down as well.
```
Analysis Summary: 
New defects found: 0 
Defects eliminated: 3
```
:+1:

**Pull request merged**